app/controllers/callbacks_controller.rb

```diff
@@ -0,0 +1,11 @@
+class CallbacksController < Devise::OmniauthCallbacksController
+    def facebook
+        @user = User.from_omniauth(request.env["omniauth.auth"])
+        sign_in_and_redirect @user
+    end
+
+    def twitter
+        @user = User.from_omniauth(request.env["omniauth.auth"])
```

Comment:
These two methods are identical. Pull them into a single method and have your facebook/twitter methods call it.